src/Snafu.md

```diff
@@ -370,6 +371,196 @@ enum Error {
+println!("{user_id:?} experienced an error");
+```
+
+### Automatically provided data
+
+By default, `source` and `backtrace` fields are exposed to the
+provider API. Additionally, any data provided by the wrapped error
+will be available on the wrapping error:
+
+```rust,ignore
+use snafu::{prelude::*, IntoError};
+
+#[derive(Debug)]
+struct UserId(u8);
+
+#[derive(Debug, Snafu)]
+struct InnerError {
+    #[snafu(provide)]
+    user_id: UserId,
+    backtrace: snafu::Backtrace,
+}
+
+#[derive(Debug, Snafu)]
+struct OuterError {
+    source: InnerError,
+}
+
+let outer = OuterSnafu.into_error(InnerSnafu { user_id: UserId(0) }.build());
+let outer = &outer as &dyn std::error::Error;
+
+// We can get the source error and downcast it at once
+outer
+    .request_ref::<InnerError>()
+    .expect("Must have a source");
+
+// We can get the deepest backtrace
+outer
+    .request_ref::<snafu::Backtrace>()
+    .expect("Must have a backtrace");
+
+// We can get arbitrary values from sources as well
+outer.request_ref::<UserId>().expect("Must have a user id");
+```
+
+## Explicitly provided data
+
+When used on a field, the `#[snafu(provide)]` attribute will expose
+that field as a reference, allowing it to be used with
+[`request_ref`][]. For more control, the `#[snafu(provide)]` attribute
+can be placed on the error struct or enum variant. In this location,
+you supply a type and an expression that will generate that type:
+
+```rust,ignore
+use snafu::prelude::*;
+
+#[derive(Debug, PartialEq)]
+struct HttpCode(u16);
+
+const HTTP_NOT_FOUND: HttpCode = HttpCode(404);
+
+#[derive(Debug, Snafu)]
+#[snafu(provide(HttpCode => HTTP_NOT_FOUND))]
+struct WebserverError;
+
+let e = WebserverError;
+let e = &e as &dyn std::error::Error;
+assert_eq!(Some(HTTP_NOT_FOUND), e.request_value::<HttpCode>());
+```
+
+The expression may access any field of the error as well as `self`:
+
+```rust,ignore
+use snafu::prelude::*;
+
+#[derive(Debug, PartialEq)]
+struct Summation(u8);
+
+#[derive(Debug, Snafu)]
+#[snafu(provide(Summation => Summation(left_side + right_side)))]
+struct AdditionError {
+    left_side: u8,
+    right_side: u8,
+}
+
+let e = AdditionSnafu {
+    left_side: 1,
+    right_side: 2,
+}
+.build();
+let e = &e as &dyn std::error::Error;
+assert_eq!(Some(Summation(3)), e.request_value::<Summation>());
+```
+
+### Configuring how data is provided
+
+You may also provide a number of optional flags that control how the
+provided data will be exposed:
+
+```rust,ignore
+use snafu::{prelude::*, IntoError};
+
+#[derive(Debug, PartialEq)]
+enum Severity {
+    Bad,
+    ReallyBad,
+}
+
+#[derive(Debug, Snafu)]
+#[snafu(provide(Severity => Severity::Bad))]
+struct InnerError;
+
+#[derive(Debug, Snafu)]
+// We want to expose `name` as a `&str`, not `&String`
+#[snafu(provide(ref, str => name))]
+// We want to expose `enabled` as a `bool`, not an `Option<bool>`
+#[snafu(provide(opt, bool => *enabled))]
```

Comment:
When the data is being provided, we get references to all of the fields. Since this example does _not_ use `ref` to denote it's returning a reference, it's returning a `bool` instead of `&bool`.